File: lab_5/1.cpp
#include <bits/stdc++.h>
using namespace std;


vector <int> graph[10000000];
int distances[10000000];
queue <int> q;

int breadthFirstSearch(int source, int size) {
    distances[source] = 0;
    q.push(source);

    int u, v;

    while (!q.empty()) {
        u = q.front();
        q.pop();

        for (int i = 0; i < graph[u].size(); i++) {
            v = graph[u][i];

            if (distances[v] == -1) {
                distances[v] = 1 + distances[u];
                q.push(v);
            }
        }
    }

    int max_distance = -1;

    for (int i = 1; i <= size; i++) {
        if (distances[i] >= max_distance) {
            max_distance = distances[i];
        }
    }

    return max_distance;
}

int main() {
    int vertices, edges;

    cin >> vertices >> edges;

    for (int i = 0; i < vertices + 1; i++) {
        distances[i] = -1;
    }

    int u, v;
    for (int i = 0; i < edges; i++) {
        cin >> u >> v;

        graph[u].push_back(v);

        // For Undiredted graphs
        if (u != v) {
            graph[v].push_back(u);
        }
    }

    int farthest_distances = breadthFirstSearch(1, vertices);

    cout << farthest_distances << endl;

    return 0;
}

File: lab_5/1.py
def breadth_first_search(grph: list, source: int):
    queue = []
    distance = [-1] * len(grph)

    distance[source] = 0
    queue.append(source)

    while len(queue) != 0:
        vertex = queue[0]
        queue.pop(0)

        i = 0
        while i < len(grph[vertex]):
            connected_vertex = grph[vertex][i]

            if connected_vertex >= len(distance):
                distance = [-1] * (connected_vertex + 1)

            if distance[connected_vertex] == -1:
                distance[connected_vertex] = 1 + distance[vertex]
                queue.append(connected_vertex)

            i += 1

    return distance


if __name__ == '__main__':

    vertices, edges = map(int, input().split())

    rows = vertices + 1

    graph = [[] for i in range(rows)]

    i = 0
    while i < edges:
        vertex, connected_vertex = map(int, input().split())
        graph[vertex].append(connected_vertex)

        if vertex != connected_vertex:
            graph[connected_vertex].append(vertex)
        i += 1

    distances = breadth_first_search(grph=graph, source=1)

    farthest_vertex = max(distances)

    print(farthest_vertex)
